pandas/core/index.py

```diff
@@ -4393,13 +4418,23 @@ def _reference_duplicate_name(self, name):
         # count the times name equals an element in self.names.
         return sum(name == n for n in self.names) > 1
 
-    def _format_native_types(self, **kwargs):
+    def _format_native_types(self, na_rep='nan', **kwargs):
         # we go through the levels and format them
-        levels = [level._format_native_types(**kwargs)
+        levels = [level._format_native_types(na_rep=na_rep, **kwargs)
                   for level in self.levels]
         mi = MultiIndex(levels=levels, labels=self.labels, names=self.names,
                         sortorder=self.sortorder, verify_integrity=False)
-        return mi.values
+        # if any index contains only NaNs, it has collapsed into an empty
+        # Float64Index, and when the multiindex has been recomposed
+        # the NaNs have come back as NaNs, not as strings corresponding to
+        # na_rep
```

Comment:
So the way to fix this is set both the levels & labels; e.g. you append a new value to the level (which will represent the nans), then change the -1 in the labels to that value (the position of that value, e.g. 1 in this case). This will then reformat the MultiIndex to work correctly.

```
In [20]: df.index.set_levels([0.0,'_'],level=0).set_labels([0,1],level=0).values
Out[20]: array([(0.0, 0), ('_', 1)], dtype=object)
```